File: backend/app/api/american_surface_atlas.py
from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.services.american_surface_atlas import (
    build_american_surface_atlas,
)


router = APIRouter(
    prefix="/research",
    tags=["research"],
)


class AmericanSurfaceAtlasRequest(BaseModel):
    spot: float = Field(
        default=100.0,
        gt=0,
    )

    strike: float = Field(
        default=100.0,
        gt=0,
    )

    rate: float = 0.05

    volatility: float = Field(
        default=0.20,
        gt=0,
    )

    maturity: float = Field(
        default=1.0,
        gt=0,
    )

    dividend_yield: float = 0.0

    s_max: float = Field(
        default=250.0,
        gt=0,
    )

    space_steps: int = Field(
        default=80,
        ge=30,
        le=200,
    )

    time_steps: int = Field(
        default=80,
        ge=30,
        le=200,
    )

    crr_steps: int = Field(
        default=150,
        ge=25,
        le=500,
    )

    crr_surface_points: int = Field(
        default=31,
        ge=15,
        le=51,
    )

    boundary_tolerance: float = Field(
        default=1e-4,
        gt=0,
    )


@router.post(
    "/american-surface-atlas"
)
def american_surface_atlas(
    request: AmericanSurfaceAtlasRequest,
):
    atlas = build_american_surface_atlas(
        spot=request.spot,
        strike=request.strike,
        rate=request.rate,
        volatility=request.volatility,
        maturity=request.maturity,
        dividend_yield=
            request.dividend_yield,
        s_max=request.s_max,
        space_steps=request.space_steps,
        time_steps=request.time_steps,
        crr_steps=request.crr_steps,
        crr_surface_points=
            request.crr_surface_points,
        boundary_tolerance=
            request.boundary_tolerance,
    )

    return {
        "option_type": "put",

        "reference_method":
            "projected_crank_nicolson",

        "input": {
            "spot": request.spot,
            "strike": request.strike,
            "rate": request.rate,
            "volatility":
                request.volatility,
            "maturity":
                request.maturity,
            "dividend_yield":
                request.dividend_yield,
            "s_max":
                request.s_max,
        },

        "grid": {
            "spot":
                atlas.spot_grid,

            "time_to_maturity":
                atlas.time_to_maturity_grid,

            "space_steps":
                atlas.space_steps,

            "time_steps":
                atlas.time_steps,

            "crr_steps":
                atlas.crr_steps,
        },

        "surfaces": {
            "crank_nicolson":
                atlas.cn_surface,

            "crr":
                atlas.crr_surface,

            "pinn_v2":
                atlas.pinn_surface,  

            "payoff":
                atlas.payoff_surface,

            "exercise_gap":
                atlas.exercise_gap_surface,
        },

        "errors": {
            "crr_signed":
                atlas.crr_signed_error_surface,

            "crr_absolute":
                atlas.crr_absolute_error_surface,

            "pinn_signed":
                atlas.pinn_signed_error_surface,

            "pinn_absolute":
                atlas.pinn_absolute_error_surface,
        },

        "pinn_convergence": {
            "available":
                atlas.pinn_convergence_available,

            "epochs":
                atlas.pinn_convergence_epochs,

            "surfaces":
                atlas.pinn_convergence_surfaces,

            "signed_errors":
                atlas.pinn_convergence_signed_errors,

            "absolute_errors":
                atlas.pinn_convergence_absolute_errors,

            "metrics":
                atlas.pinn_convergence_metrics,

            "improvement_surface":
                atlas.pinn_improvement_surface,

            "boundary_diagnostics":
                atlas.pinn_boundary_diagnostics,

            "boundary_distance_profiles":
                atlas.pinn_boundary_distance_profiles,
        },

        "exercise_boundary":
            atlas.exercise_boundary,

        "pinn": {
            "available":
                atlas.pinn_available,

            "method":
                atlas.pinn_method,

            "final_loss":
                atlas.pinn_final_loss,

            "training_seconds":
                atlas.pinn_training_seconds,

            "mae_vs_cn":
                atlas.pinn_mae,

            "rmse_vs_cn":
                atlas.pinn_rmse,

            "max_absolute_error_vs_cn":
                atlas.pinn_max_absolute_error,
        },

        "summary": {
            "min_price":
                atlas.min_price,

            "max_price":
                atlas.max_price,

            "max_exercise_gap":
                atlas.max_exercise_gap,

            "max_crr_absolute_error":
                atlas.max_crr_absolute_error,
        },
    }

File: frontend/lib/api/americanSurfaceAtlas.ts
export type ExerciseBoundaryPoint = {
  time_to_maturity: number;
  spot: number | null;
};

export type AmericanSurfaceAtlasInput = {
  spot: number;
  strike: number;
  rate: number;
  volatility: number;
  maturity: number;
  dividend_yield: number;

  s_max?: number;
  space_steps?: number;
  time_steps?: number;

  crr_steps?: number;
  crr_surface_points?: number;

  boundary_tolerance?: number;
};

export type AmericanSurfaceAtlasResult = {
  option_type: "put";

  reference_method:
    "projected_crank_nicolson";

  input: {
    spot: number;
    strike: number;
    rate: number;
    volatility: number;
    maturity: number;
    dividend_yield: number;
    s_max: number;
  };

  grid: {
    spot: number[];
    time_to_maturity: number[];

    space_steps: number;
    time_steps: number;
    crr_steps: number;
  };

  surfaces: {
    crank_nicolson: number[][];
    crr: number[][];
    pinn_v2: number[][] | null;
    payoff: number[][];
    exercise_gap: number[][];
  };

  errors: {
    crr_signed: number[][];
    crr_absolute: number[][];
    pinn_signed: number[][] | null;
    pinn_absolute: number[][] | null;
  };

  pinn: {
    available: boolean;
    method: string | null;
    final_loss: number | null;
    training_seconds: number | null;
    mae_vs_cn: number | null;
    rmse_vs_cn: number | null;
    max_absolute_error_vs_cn:
      number | null;
  };

  pinn_convergence: {
    available: boolean;

    epochs: number[];

    surfaces: Record<
        string,
        number[][]
    >;

    signed_errors: Record<
        string,
        number[][]
    >;

    absolute_errors: Record<
        string,
        number[][]
    >;

    metrics: Record<
        string,
        {
        mae: number;
        rmse: number;
        max_absolute_error: number;
        training_loss: number;
        elapsed_training_seconds: number;
        inference_seconds: number;
        }
    >;

    improvement_surface:
        number[][] | null;

    boundary_diagnostics: Record<
        string,
        {
            band_width: number;

            near_boundary: {
            count: number;
            mae: number | null;
            rmse: number | null;
            max_absolute_error: number | null;
            };

            away_from_boundary: {
            count: number;
            mae: number | null;
            rmse: number | null;
            max_absolute_error: number | null;
            };

            near_to_away_mae_ratio:
            number | null;

            worst_near_boundary: {
            spot: number;
            time_to_maturity: number;
            boundary_spot: number;
            distance_to_boundary: number;
            absolute_error: number;
            } | null;

            worst_away_from_boundary: {
            spot: number;
            time_to_maturity: number;
            boundary_spot: number;
            distance_to_boundary: number;
            absolute_error: number;
            } | null;
        }
        >;

    boundary_distance_profiles: Record<
        string,
        {
            distance_definition: string;
            grid_spacing: number;
            strike: number;
            observation_count: number;
            distance_error_correlation: number | null;

            bins: Array<{
            label: string;
            lower_distance: number;
            upper_distance: number | null;
            count: number;
            mae: number | null;
            rmse: number | null;
            median_absolute_error: number | null;
            p90_absolute_error: number | null;
            max_absolute_error: number | null;
            }>;
        }
        >;
    };

  exercise_boundary:
    ExerciseBoundaryPoint[];

  summary: {
    min_price: number;
    max_price: number;
    max_exercise_gap: number;
    max_crr_absolute_error: number;
  };
};


const API_URL =
  process.env.NEXT_PUBLIC_API_URL ??
  "http://127.0.0.1:8000";


export async function getAmericanSurfaceAtlas(
  input: AmericanSurfaceAtlasInput
): Promise<AmericanSurfaceAtlasResult> {
  const response = await fetch(
    `${API_URL}/research/american-surface-atlas`,
    {
      method: "POST",

      headers: {
        "Content-Type":
          "application/json",
      },

      body: JSON.stringify(input),
    }
  );

  if (!response.ok) {
    const text =
      await response.text();

    throw new Error(
      `American Surface Atlas API error: ${
        response.status
      } ${text}`
    );
  }

  return response.json();
}